packages/actor-extract-links-extract-tree/lib/ActorExtractLinksTree.ts

```diff
@@ -42,21 +41,45 @@ export class ActorExtractLinksTree extends ActorExtractLinks {
         this.getTreeQuadsRawRelations(quad,
           currentNodeUrl,
           pageRelationNodes,
-          nodeLinks));
+          nodeLinks,
+          relationDescriptions));
 
       // Resolve to discovered links
-      metadata.on('end', () => {
+      metadata.on('end', async() => {
         // Validate if the node forward have the current node as implicit subject
         for (const [ nodeValue, link ] of nodeLinks) {
           if (pageRelationNodes.has(nodeValue)) {
-            links.push({ url: link });
+            const relationDescription = relationDescriptions.get(nodeValue);
+            if (typeof relationDescription !== 'undefined') {
+              relations.push(collectRelation(relationDescription, link));
+            } else {
+              relations.push(collectRelation({}, link));
+            }
           }
         }
-        resolve({ links });
+
+        const node: ITreeNode = { relation: relations, subject: currentNodeUrl };
+        let acceptedRelation = relations;
+
+        const filters = await this.applyFilter(node, action.context);
+        acceptedRelation = this.handleFilter(filters, acceptedRelation);
+        resolve({ links: acceptedRelation.map(el => ({ url: el.node })) });
       });
     });
   }
 
+  /* istanbul ignore next */
```

Comment:
I ignore this function because it only exist so I can mock the filter class. If I were to test this function I would not be unit test because I would be testing two units.

Reply:
But why should it be mocked? Can't you test directly using the `FilterNode` implementation?

Reply:
I think it's better to test it separately, as the two component does some pretty different things. I don't think those two test should depend on each other. It also make the test more verbose.

Reply:
I think with another architecture the `FilterNode` should be injected into the actor.

Reply:
I can do it but for sure, but at the same time we are not doing it when it's a mediator.

Reply:
> we are not doing it when it's a mediator.

We are though, that's what the integration and spec tests are for.
(but they probably don't have 100% coverage on that level)

Reply:
But I thought we didn't have integration test inside this version of comunica.

Reply:
yes, I agree!